experimental/GroebnerWalk/src/generic_walk.jl

```diff
@@ -0,0 +1,245 @@
+
+  @v_do :groebner_walk steps = 0
+  while !isempty(v)
+    @vprintln :groebner_walk v
+    MG = generic_step(MG, v, target)
+    v = next_gamma(MG, v, start, target)
+
+    @v_do :groebner_walk steps += 1
+    @vprintln :groebner_walk 2 G
+    @vprintln :groebner_walk 2 "======="
+  end
+
+  @vprint :groebner_walk "Cones crossed: "
+  @vprintln :groebner_walk steps
+  return gens(MG)
+end
+
+@doc raw"""
+    generic_step(MG::MarkedGroebnerBasis, v::Vector{ZZRingElem}, ord::MonomialOrdering)
+
+Given the marked Gröbner basis `MG` and a facet normal vector `v`, compute the next marked Gröbner basis.
+"""
+function generic_step(MG::MarkedGroebnerBasis, v::Vector{ZZRingElem}, ord::MonomialOrdering)
+  facet_generators = facet_initials(MG, v)
+  H = groebner_basis(
+    ideal(facet_generators); ordering=ord, complete_reduction=true, algorithm=:buchberger
+  )
+
+  @vprint :groebner_walk 5 "Initials forms of facet: "
+  @vprintln :groebner_walk 5 facet_generators
+
+  @vprint :groebner_walk 3 "GB of initial forms: "
+  @vprintln :groebner_walk 3 H
+
+  H = MarkedGroebnerBasis(gens(H), leading_term.(H; ordering = ord))
+
+  lift_generic!(H, MG)
+  autoreduce!(H)
+
+  return H
+end
+
+@doc raw"""
+    difference_lead_tail(MG::MarkedGroebnerBasis)
+
+Computes $a - b$ for $a$ a leading exponent and $b$ in the tail of some $g\in MG$. 
+"""
+function difference_lead_tail(MG::MarkedGroebnerBasis)
+  (G,Lm) = gens(MG), markings(MG)
+  lead_exp = leading_exponent_vector.(Lm)
+  
+  v = map(zip(lead_exp, exponents.(G))) do (l,T)
+    Ref(l) .- T
+  end
+  
+  return [ZZ.(v)./ZZ(gcd(v)) for v in unique!(reduce(vcat, v)) if !iszero(v)]
+end
+
+@doc raw"""
+    next_gamma(
+      MG::MarkedGroebnerBasis, w::Vector{ZZRingElem}, start::MonomialOrdering, target::MonomialOrdering
+    )
+
+Given a marked Gröbner basis `MG`, a weight vector `w` and monomial orderings `start` and `target`,
+returns the "next" facet normal, i.e. the bounding vector $v$ fulfilling $w<v$ and being minimal with respect to the facet preorder $<$.
+"""
+function next_gamma(
+    MG::MarkedGroebnerBasis, w::Vector{ZZRingElem}, start::MonomialOrdering, target::MonomialOrdering
+  )
+    V = filter_by_ordering(start, target, difference_lead_tail(MG))
+    if w != ZZ.([0]) 
+        V = filter_lf(w, start, target, V)
+    end
+    if isempty(V)
+        return V
+    end
+
+    return reduce(V[2:end], init=first(V)) do v,w
+      if facet_less_than(canonical_matrix(start), canonical_matrix(target), v, w)
+          return v
+      else 
+          return w
+      end
+  end
+end
+
+@doc raw"""
+    facet_initials(MG::MarkedGroebnerBasis, v::Vector{ZZRingElem})
+
+Given a marked Gröbner basis `MG` and a facet normal `v`, computes the Gröbner basis of initial forms 
+of `MG` by truncating to all bounding vectors parallel to `v`.
+"""
+function facet_initials(MG::MarkedGroebnerBasis, v::Vector{ZZRingElem})
+  R = base_ring(MG)
+  inwG = Vector{MPolyRingElem}()
```

Comment:
```suggestion
  inwG = elem_type(R)[]
```
this initializes the vector with the concrete mpolynomial type, instead of using the abstract mpoly type (that doesn't know about it's coefficient type)

Reply:
This should also be changed in `autoreduce()`.